content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,82 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+
+
+## Reference Solution
+
+Some test suites require a reference solution to generate the expected value(s) and compare them to those returned by the user solution. While such practice is common and sometimes necessary, there are things which have to be handled carefully:
+
+- **Avoid using a reference solution at all if possible.** Sometimes there's no other way, but quite often a reference solution is simply not necessary: test cases can be generated with the answer known upfront, which eliminates an entire class of potential problems associated with the use of reference solutions:
+  - Extra time spent on running a reference solution
+  - Reference solution being accessible to users by mistake
+  - Input mutation by the user solution affecting the input passed to the reference solution
+  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
+- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function.
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs and outputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s). Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
+- **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
```

Comment:
`null` usually isn't like an empty array or an empty string. For most problems `null` is a language-specific implementation detail rather than a value that makes sense, so there's no reasonable result to return, for example, if a `String` is `null`. I assume that if there isn't valid output for some input, there won't be such input. Whether it should or doesn't have to be specified explicitly in such cases, `null` tests often don't make sense.

Reply:
How do you suggest it should be handled then? Should `null` inputs be considered a design issue with a task?

Reply:
Actually, this looks correct overall, but I guess the question here is what "explicitly" is. If it says "you are given a string", it already means that it can't be `null` because a "`string`" (as a data type) may be `null` in some language, but a "string" (as a normal word) can't.
Another thing is that this sentence seems to encourage testing `null` input, while it's not normal in most cases.

Reply:
```suggestion
- **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. If description states that solution should handle some specific inputs, like for example empty arrays, empty strings, etc. then a corresponding fixed test case should be added.
```

Do you think it's any better?

Reply:
There's no need to mention empty arrays and strings in the description if they aren't exceptions that should be handled in a different way.
* Find the sum of the elements => nothing special about an empty array here, it should be tested.
* Find the shortest string => if the opposite isn't clear from return types in all available languages, there's no possible unique answer for an empty collection, so it's not tested by default.

Reply:
How would you describe here requirement of "returm minimum number from array, or - 1 if empty", in context of this paragraph? Should it be simply not allowed as a crappy requirement?

Reply:
It doesn't make sense, so it's a bad requirement in general. It may be acceptable in lower level languages like C if the input is unsigned and when alternative ways have been analyzed. In more complicated cases it might be acceptable in a kata where the main focus is on math and calculations and where coding style is secondary, but even then that's not ideal.

Reply:
> invalid inputs and edge cases if the problem allows them

Invalid input is something that's not expected by default, edge cases are just a special subset of normal input, so I'm not sure about grouping them together like this...

Reply:
So you mean that since description mentions them, it's not invalid? Yeah, makes sense. 👍

Reply:
Now it looks like "edge cases" are opposed to "valid inputs". But it's like: "you are given a number (signed integer), make sure it's non-negative and check if it's a perfect square; if the number is negative, throw an InvalidArgumentException". Non-negative numbers are valid input; negative numbers probably are invalid because the possibility of passing a negative number probably is an implementation detail, but since they can be passed, they may be called valid, it's kind of ambiguous here; 0 is an edge case (and maybe 1 as well), but it's definitely valid.

Reply:
Yeah you are right, and it seems that my English is too poor to word this paragraph properly :( Is there anything you'd propose? Remove the whole bullet point, reword it, or replace "valid" with something else, like "expected inputs", "normal/regular inputs", idk? I am all ears :)

Reply:
I've edited it, but I still don't like how it sounds. Edge cases probably shouldn't be grouped with special cases in one sentence.

Reply:
To me, it sounds really cool. Thanks!